File: apps/worker/src/processors/enrich.ts
import type { Job } from 'bullmq'
import * as cheerio from 'cheerio'
import { getDb, getJobById, updateJob, logEvent } from '@jobpilot/db'
import { randomUserAgent, randomDelay, DELAY_PAGE } from '@jobpilot/anti-detection'

function extractDescription(html: string): string | null {
  const $ = cheerio.load(html)

  // Strategy 1: JSON-LD JobPosting
  const ldScripts = $('script[type="application/ld+json"]')
  for (let i = 0; i < ldScripts.length; i++) {
    try {
      let data = JSON.parse($(ldScripts[i]).text() || '{}')
      if (Array.isArray(data)) {
        data = data.find((d: Record<string, unknown>) => d['@type'] === 'JobPosting')
      }
      if (data?.['@type'] === 'JobPosting' && data.description) {
        const desc = cheerio.load(data.description as string).text().trim()
        if (desc) return desc
      }
    } catch {
      // ignore parse errors
    }
  }

  // Strategy 2: meta description
  const meta = $('meta[name="description"]').attr('content')?.trim()
  if (meta && meta.length > 100) return meta

  // Strategy 3: common job description containers
  const selectors = [
    '#jobDescriptionText',
    '.jobsearch-jobDescriptionText',
    '[class*="description"]',
    '[class*="jobDescription"]',
    'article',
    'main',
  ]
  for (const sel of selectors) {
    const text = $(sel).first().text().trim()
    if (text.length > 100) return text
  }

  // Fallback: body text
  const body = $('body').text().trim()
  if (body.length > 100) return body.slice(0, 5000)

  return null
}

export async function processEnrich(job: Job<{ jobId: number }>): Promise<{ jobIds: number[] }> {
  const db = getDb()
  const dbJob = getJobById(db, job.data.jobId)
  if (!dbJob) return { jobIds: [] }

  console.log(`[enrich] Buscando: ${dbJob.url}`)

  try {
    const resp = await fetch(dbJob.url, {
      headers: { 'User-Agent': randomUserAgent() },
      redirect: 'follow',
      signal: AbortSignal.timeout(15_000),
    })
    if (!resp.ok) {
      logEvent(db, dbJob.id, 'enrich', 'warning', `Falha na busca: ${resp.status}`)
      return { jobIds: [] }
    }

    const html = await resp.text()
    const description = extractDescription(html)

    if (description) {
      updateJob(db, dbJob.id, { description, status: 'enriched' })
      logEvent(db, dbJob.id, 'enrich', 'info', `Enriquecida (${description.length} caracteres)`)
    } else {
      // Move to enriched anyway so pipeline continues
      updateJob(db, dbJob.id, { status: 'enriched' })
      logEvent(db, dbJob.id, 'enrich', 'warning', 'Nenhuma descrição extraída')
    }

    await randomDelay(...DELAY_PAGE)
    return { jobIds: [dbJob.id] }
  } catch (err) {
    logEvent(db, dbJob.id, 'enrich', 'warning', `Erro na busca: ${err}`)
    return { jobIds: [] }
  }
}

File: apps/agent/pipeline/enrich.py
"""Stage 2: Enrich job listings with full descriptions."""

from __future__ import annotations

import logging
import sqlite3

import requests
from bs4 import BeautifulSoup

from apps.agent.anti_detection import random_delay, random_user_agent, DELAY_PAGE
from apps.agent.db import get_jobs, update_job, log_event

logger = logging.getLogger(__name__)


def _extract_description(html: str) -> str | None:
    """Extract job description from HTML using a cascade of strategies."""
    soup = BeautifulSoup(html, "html.parser")

    # Strategy 1: JSON-LD JobPosting
    for script in soup.find_all("script", type="application/ld+json"):
        try:
            import json
            data = json.loads(script.string or "")
            if isinstance(data, list):
                data = next((d for d in data if d.get("@type") == "JobPosting"), None)
            if data and data.get("@type") == "JobPosting":
                desc = data.get("description", "")
                if desc:
                    return BeautifulSoup(desc, "html.parser").get_text(separator="\n", strip=True)
        except (json.JSONDecodeError, StopIteration):
            continue

    # Strategy 2: meta description
    meta = soup.find("meta", attrs={"name": "description"})
    if meta and meta.get("content", "").strip():
        content = meta["content"].strip()
        if len(content) > 100:
            return content

    # Strategy 3: main content heuristic — look for common job description containers
    for selector in [
        "#jobDescriptionText",
        ".jobsearch-jobDescriptionText",
        '[class*="description"]',
        '[class*="jobDescription"]',
        "article",
        "main",
    ]:
        el = soup.select_one(selector)
        if el:
            text = el.get_text(separator="\n", strip=True)
            if len(text) > 100:
                return text

    # Fallback: body text
    body = soup.find("body")
    if body:
        text = body.get_text(separator="\n", strip=True)
        if len(text) > 100:
            return text[:5000]

    return None


def run(db: sqlite3.Connection) -> int:
    """Enrich jobs that lack descriptions. Returns count of enriched jobs."""
    jobs = get_jobs(db, status="discovered")
    enriched = 0

    for job in jobs:
        url = job["url"]
        logger.info("Enriching: %s", url)

        try:
            resp = requests.get(
                url,
                headers={"User-Agent": random_user_agent()},
                timeout=15,
                allow_redirects=True,
            )
            resp.raise_for_status()
        except requests.RequestException as e:
            logger.warning("Failed to fetch %s: %s", url, e)
            log_event(db, job["id"], "enrich", "warning", f"Fetch failed: {e}")
            continue

        description = _extract_description(resp.text)
        if description:
            update_job(db, job["id"], description=description, status="enriched")
            log_event(db, job["id"], "enrich", "info", f"Enriched ({len(description)} chars)")
            enriched += 1
        else:
            logger.warning("No description found for %s", url)
            log_event(db, job["id"], "enrich", "warning", "No description extracted")
            # Move to enriched anyway so pipeline continues
            update_job(db, job["id"], status="enriched")

        random_delay(*DELAY_PAGE)

    return enriched
